### PR TITLE
Add FiftyOne datamodule

### PR DESCRIPTION
**Pull request by lumurillo**
# What does this PR do?

Add a new datamodule that loads FiftyOne datasets. This implementation let MART delegate the data handling to FiftyOne, where the user can do data curation tasks.

This data handling would be agnostic to MART, because at the end here is adapted a `torchvision` dataset.

I set this implementation to be optional. The user can install the fiftyone dependencies by running:
```shell
pip install -e .[fiftyone]
```

**Important**: 
* To create a dataset in FiftyOne, it's recommended to use the CLI feature: https://docs.voxel51.com/cli/index.html#create-datasets
* To launch the FiftyOne app: https://docs.voxel51.com/cli/index.html#launch-the-app

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/configs/datamodule/fiftyone.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  dataset_name: ???
+  gt_field: ${..train_dataset.gt_field}
+  sample_tags: []
+  label_tags: []
+  transforms:
+    _target_: mart.transforms.Compose
+    transforms:
+      - _target_: torchvision.transforms.ToTensor
+      - _target_: mart.transforms.ConvertCocoPolysToMask
+      - _target_: mart.transforms.RandomHorizontalFlip
```
Comment (mzweilin) [inline mart/configs/datamodule/fiftyone.yaml:63]: We usually don't perform random transformation in validation/test.

File: mart/configs/datamodule/fiftyone.yaml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
```
Comment (mzweilin) [inline mart/configs/datamodule/fiftyone.yaml:39]: We usually don't perform random transformation in validation/test.

**Review by mzweilin — CHANGES_REQUESTED** [review]
Thanks for bringing the new feature. I was able to install, but I don't have an example to test with.

File: mart/datamodules/fiftyone.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    def __getitem__(self, index: int) -> Any:
+        sample_id = self.ids[index]
+        sample = self.filtered_dataset[sample_id]
+        metadata = sample.metadata
+        img = Image.open(sample.filepath).convert("RGB")
+
+        target = {}
+        target["image_id"] = index
+        target["file_name"] = sample.filepath
```
Comment (mzweilin) [inline mart/datamodules/fiftyone.py:80]: We just need the file name, not the full path here `os.path.basename(sample.filepath)`

File: mart/datamodules/fiftyone.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        **kwargs,
+    ) -> None:
+        super().__init__("", transforms, transform, target_transform)
+
+        self.gt_field = gt_field
+
+        # load FiftyOne dataset
+        if dataset_name in fo.list_datasets():
+            self.dataset = fo.load_dataset(dataset_name)
+        elif dataset_name in foz.list_zoo_datasets():
```
Comment (mzweilin) [inline mart/datamodules/fiftyone.py:50]: I don't think we want to load zoo datasets here, because it tries to index the dataset every time.

Maybe we should revert the changes, and use the FiftyOne command line to pre-load zoo datasets as datasets. For example, `fiftyone zoo datasets load -s validation -n coco-2017-validation coco-2017` loads the validation split of coco-2017 as a dataset named "coco-2017-validation".

**Comment by mzweilin** [comment]
# FiftyOne commands to load (index) datasets.

Use COCO-2017 as an example. Unfortunately, FiftyOne does not support person-keypoints annotations yet.

## Download and load zoo datasets

```bash
fiftyone zoo datasets load \
coco-2017 \
-s train \
-n coco-2017-instances-train \
-k include_id=true label_types=detections,segmentations

fiftyone zoo datasets load \
coco-2017 \
-s validation \
-n coco-2017-instances-validation \
-k include_id=true label_types=detections,segmentations
```

## Load local datasets

```bash
fiftyone datasets create \
--name coco-2017-instances-validation \
--dataset-dir /raid/datasets/coco/ \
--type fiftyone.types.COCODetectionDataset \
--kwargs \
data_path="val2017" \
labels_path=/raid/datasets/coco/annotations/instances_val2017.json \
persistent=true \
include_id=true
```

## Use the FiftyOne datamodule

```yaml
datamodule:
  train_dataset:
    dataset_name: coco-2017-instances-train
    gt_field: segmentations
  val_dataset:
    dataset_name: coco-2017-instances-validation
    gt_field: segmentations
```

**Comment by lumurillo** [comment]
> # FiftyOne commands to load (index) datasets.
> Use COCO-2017 as an example. Unfortunately, FiftyOne does not support person-keypoints annotations yet.
> 
> ## Download and load zoo datasets
> ```shell
> fiftyone zoo datasets load \
> coco-2017 \
> -s train \
> -n coco-2017-instances-train \
> -k include_id=true label_types=detections,segmentations
> 
> fiftyone zoo datasets load \
> coco-2017 \
> -s validation \
> -n coco-2017-instances-validation \
> -k include_id=true label_types=detections,segmentations
> ```
> 
> ## Load local datasets
> ```shell
> fiftyone datasets create \
> --name coco-2017-instances-validation \
> --dataset-dir /raid/datasets/coco/ \
> --type fiftyone.types.COCODetectionDataset \
> --kwargs \
> data_path="val2017" \
> labels_path=/raid/datasets/coco/annotations/instances_val2017.json \
> persistent=true \
> include_id=true
> ```
> 
> ## Use the FiftyOne datamodule
> ```yaml
> datamodule:
>   train_dataset:
>     dataset_name: coco-2017-instances-train
>     gt_field: segmentations
>   val_dataset:
>     dataset_name: coco-2017-instances-validation
>     gt_field: segmentations
> ```

Is this for the for the example documentation?

File: mart/datamodules/fiftyone.py (hunk 4 above)
Comment (lumurillo) [reply mart/datamodules/fiftyone.py:50]: ok, makes sense
